packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -1860,6 +1889,72 @@ export class SnapController extends BaseController<
     return JSON.stringify(encryptedState);
   }
 
+  /**
+   * Get the new Snap state to persist based on the given state and encryption
+   * flag.
+   *
+   * - If the state is null, return null.
+   * - If the state should not be encrypted, return the JSON stringified state.
+   * - Otherwise if the state should be encrypted, return the encrypted state.
+   *
+   * @param snapId - The Snap ID.
+   * @param state - The state to persist.
+   * @param encrypted - A flag to indicate whether to use encrypted storage or
+   * not.
+   * @returns The state to persist.
+   */
+  async #getStateToPersist(
+    snapId: SnapId,
+    state: Record<string, Json> | null,
+    encrypted: boolean,
+  ) {
+    if (state === null) {
+      return null;
+    }
+
+    if (encrypted) {
+      return await this.#encryptSnapState(snapId, state);
+    }
+
+    return JSON.stringify(state);
+  }
+
+  /**
+   * Persist the state of a Snap.
+   *
+   * This is run with a mutex to ensure that only one state update per Snap is
+   * processed at a time, avoiding possible race conditions.
+   *
+   * @param snapId - The Snap ID.
+   * @param newSnapState - The new state of the Snap.
+   * @param encrypted - A flag to indicate whether to use encrypted storage or
+   * not.
+   */
+  async #persistSnapState(
+    snapId: SnapId,
+    newSnapState: Record<string, Json> | null,
+    encrypted: boolean,
+  ) {
+    const runtime = this.#getRuntimeExpect(snapId);
+    await runtime.stateMutex.runExclusive(async () => {
```

Comment:
Technically we would probably only need a mutex for encrypted storage since regular storage is synchronous? Do you think it is preferable to have both operations under the same mutex?